server/src/main/java/org/opensearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java

```diff
@@ -261,24 +262,43 @@ private void onFailure(ShardRouting shardRouting, Exception e) {
             tryNext(e, false);
         }
 
-        private ShardRouting nextRoutingOrNull() {
+        private ShardRouting nextRoutingOrNull(Exception failure) {
             if (shardsIt.size() == 0 || shardIndex >= shardsIt.size()) {
                 return null;
             }
             ShardRouting next = shardsIt.get(shardIndex).nextOrNull();
+
+            // This checks if the shard is present in data node with weighted routing weight set to 0,
+            // In such cases we fail open, if shard search request for the shard from other shard copies fail with non
+            // retryable exception.
+            while (next != null && WeightedRoutingHelper.shardInWeighedAwayAZ(next.currentNodeId(), clusterService.state())) {
+                if (WeightedRoutingHelper.isInternalFailure(failure)) {
```

Comment:
I strongly feel we should add allow searches in weighed away nodes, when there are unassigned copies of that shard as well . This would prevent unnecessary 4XX to the user .

Reply:
Yeah makes sense. I can add that check

Reply:
Lets avoid redundant handling of logic